examples/regression/Main.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE FunctionalDependencies #-}
+
+module Main where
+
+import Control.Monad (foldM)
+
+import Torch.Tensor (Tensor)
+import Torch.DType (DType (Float))
+import Torch.TensorFactories (ones', rand', randn')
+import Torch.Functions
+import Torch.Autograd
+import Torch.NN
+
+{- Types -}
+
+data LinearSpec = LinearSpec { in_features :: Int, out_features :: Int }
+  deriving (Show, Eq)
+
+data Linear = Linear { weight :: Parameter, bias :: Parameter } deriving Show
+
+{- Instances -}
+
+instance Randomizable LinearSpec Linear where
+  sample LinearSpec{..} = do
+      w <- makeIndependent =<< randn' [in_features, out_features]
+      b <- makeIndependent =<< randn' [out_features]
+      return $ Linear w b
+
+instance Parameterized Linear where
+  flattenParameters Linear{..} = [weight, bias]
+  replaceOwnParameters _ = do
+    weight <- nextParameter
+    bias <- nextParameter
+    return $ Linear{..}
+
+{- Forward functions -}
+
+linear :: Linear -> Tensor -> Tensor
+linear Linear{..} input = (matmul input dWeight) + dBias
+  where
+    (dWeight, dBias) = (toDependent weight, toDependent bias)
+  
+{- Optimization -}
+
+sgd :: Tensor -> [Parameter] -> [Tensor] -> [Tensor]
+sgd lr parameters gradients = zipWith step depParameters gradients
```

Comment:
We should move this and the `Linear` into `hasktorch/`, because right now we just keep duplicating them unnecessarily.

Reply:
For now I've moving them into NN.hs with the idea eventually refactor it splitting into optimizer / layer / typed / untyped as those APIs start to come together.